### PR TITLE
Fix training difficulty for DenseNetFCN

### PR DESCRIPTION
**Pull request by titu1994**
It was brought to my attention, that there was a connection missing in the final layer of the DenseNetFCN, which was the reason for the poor performance on other datasets. (https://github.com/titu1994/DenseNet/pull/15)

This has been tested on Keras 2.0.4+, and I can confirm that the model is performing far better than before on a private medical dataset. 

Fixes https://github.com/farizrahman4u/keras-contrib/issues/63
@ahundt, this may be of use to you.

### CODE REVIEW

**Comment by titu1994** [comment]
@farizrahman4u Could you review this PR and merge?

**Comment by ahundt** [comment]
wow great! let me test before we say #63 is fully fixed

**Comment by titu1994** [comment]
Yup that's true. But when I tested it on a private dataset it trained properly and didn't get stuck as before. 

Even considering that the number of parameters exactly matches the ones in the paper is a sign of hope.

**Comment by titu1994** [comment]
@farizrahman4u @the-moliver Could you review and merge please?

**Comment by ahundt** [comment]
I'm still training, it is just about to pass the accuracy threshold that can be reached with labeling every pixel background:
```
Epoch 56/250
 6098/11127 [===============>..............] - ETA: 2170s - loss: 2.3599 - sparse_accuracy_ignoring_last_label: 0.7199 
```

I think this is a good sign!

**Comment by titu1994** [comment]
Mine finished training, finally getting better results than the UNet baseline that I was using. 

On a side note, are you training that model on a CPU? Or does it take that long even on a GPU/s? If it takes that long on a GPU, it must be one of the largest models I've ever seen.

**Comment by ahundt** [comment]
It is on a gpu, I'm not sure why it has been so slow... While it is the largest of the models from the tiramisu paper, it is not actually that large. It might be obvious if I profile it lol.

File: keras_contrib/applications/densenet.py [hunk 1]
```diff
@@ -178,7 +178,7 @@ def DenseNet(input_shape=None, depth=40, nb_dense_block=3, growth_rate=12, nb_fi
 def DenseNetFCN(input_shape, nb_dense_block=5, growth_rate=16, nb_layers_per_block=4,
                 reduction=0.0, dropout_rate=0.0, weight_decay=1E-4, init_conv_filters=48,
                 include_top=True, weights=None, input_tensor=None, classes=1, activation='softmax',
-                upsampling_conv=128, upsampling_type='upsampling'):
+                upsampling_conv=128, upsampling_type='deconv'):
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: Because it matches the paper. Also, I'm getting slight better results with it than with upsampling or subpixel.

File: keras_contrib/applications/densenet.py [hunk 2]
```diff
@@ -382,10 +382,10 @@ def __dense_block(x, nb_layers, nb_filter, growth_rate, bottleneck=False, dropou
     x_list = [x]
 
     for i in range(nb_layers):
-        x = __conv_block(x, growth_rate, bottleneck, dropout_rate, weight_decay)
-        x_list.append(x)
+        conv_block = __conv_block(x, growth_rate, bottleneck, dropout_rate, weight_decay)
+        x_list.append(conv_block)
 
-        x = concatenate(x_list, axis=concat_axis)
+        x = concatenate([x, conv_block], axis=concat_axis)
```
Comment (titu1994) [inline keras_contrib/applications/densenet.py]: This is the important fix that I mentioned. With it, and Deconvolution upsampling, the parameter count exactly matches the paper and trains properly. 

The previous version had fewer parameters, and did not train properly.

Comment (ahundt) [inline keras_contrib/applications/densenet.py]: My version is ever so slightly different from both of your versions. I think my version may be slightly more accurate w.r.t. the paper's description. That said, it is also possible that this change of yours could also outperform the original paper. :-)

I looked carefully at the following before writing the code above:
- your previous version
- your current version
- the paper and its [corresponding code](https://github.com/SimJeg/FC-DenseNet/blob/master/FC-DenseNet.py#L67)

Note that in the version I posted `x_list` and `return_concat_list` are gone. I may very well still be mistaken here, but could you either correct me or confirm my suspicion?

Comment (titu1994) [inline keras_contrib/applications/densenet.py]: Sorry for the late reply. 

Without the return of concat list, the upsample path will not be correct anymore. As in, the upsample block concatenates a long term dependency, and that is provided by the concat_list.

I tried modifying the code to work without that, but there are fewer number of parameters than the original paper and the current implementation, and that suggests that it may suffer during training.

Comment (ahundt) [inline keras_contrib/applications/densenet.py]: Okay that's evidence enough for me! As a side test, I could try running those changes if you still have them somewhere.

Comment (titu1994) [inline keras_contrib/applications/densenet.py]: Ah I reverted my code of the changes after I was done. Sorry about that.

Comment (ahundt) [inline keras_contrib/applications/densenet.py]: oh well, think it is worthwhile?

**Comment by ahundt** [comment]
@farizrahman4u can this be merged?

**Comment by titu1994** [comment]
@farizrahman4u @the-moliver Could you merge this please? 

I've finished with MobileNets and ResNeXt, but I can't add PRs until this one gets merged (I made changes to master in my branch, and new branches would include this PR as well)
